### PR TITLE
feat: fetch and parse full gmail message

### PR DESCRIPTION
**Pull request by rostaklein**
first part of https://github.com/twentyhq/core-team-issues/issues/34
related PR https://github.com/twentyhq/twenty/pull/5081

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -33,6 +33,23 @@
       "internalConsoleOptions": "openOnSessionStart",
       "console": "internalConsole",
       "cwd": "${workspaceFolder}/packages/twenty-server/"
+    },
+    {
+      "name": "twenty-server - gmail fetch debug",
```
Comment (charlesBochet) [inline .vscode/launch.json:38]: this seems a bit overkill, let's remove it :)

File: packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts [hunk 2]
```diff
@@ -1,12 +1,6 @@
...
-  threadId: string;
-  labelIds: string[];
-  snippet: string;
-  sizeEstimate: number;
-  raw: string;
-  historyId: string;
-  internalDate: string;
+import { gmail_v1 } from 'googleapis';
+
+export type GmailMessageParsedResponse = gmail_v1.Schema$Message & {
```
Comment (charlesBochet) [inline packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts:3]: I'm not a big fan of having errors mixed with the message. Also, it seems that the new type is having all fields as optionals. Can we infer a type from gmail_v1.Schema$Message with all fields we need required?

Do we need to have errors here or can we add it as an additional parameter of functions when needed?

File: .vscode/launch.json (hunk 1 above)
Comment (rostaklein) [reply .vscode/launch.json:38]: Ah, I didnt mean to commit, but I needed a way to debug the email fetching, I was hoping that the worker would do that, but it didnt, any tips how to do that? :)

File: packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts (hunk 2 above)
Comment (rostaklein) [reply packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts:3]: Not sure, it was there before, I assume perhaps because of the way how `fetch-by-batch.service.ts` is implented? 🤔

File: .vscode/launch.json (hunk 1 above)
Comment (charlesBochet) [reply .vscode/launch.json:38]: the worker should do it actually, if you start the worker with the debugger and place a breakpoint it should respect it

File: packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts (hunk 2 above)
Comment (charlesBochet) [reply packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts:3]: aren't you adding the errors to this type?

Comment (rostaklein) [reply packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts:3]: >Do we need to have errors here or can we add it as an additional parameter of functions when needed?

I tried to have a look at this once again and even though im still not completely able to understand what `fetch-by-batch.service.ts` it apparently adds the `error` property to the parsed object when parsing the response here https://github.com/twentyhq/twenty/blob/main/packages/twenty-server/src/modules/messaging/services/fetch-by-batch/fetch-by-batch.service.ts#L107

We can of course make the type nicer by making it a literal like `gmail_v1.Schema$Message | GmailError` but just imagine looping through such an array. Currently it checks whether the `error` property exists and returns early if it does https://github.com/twentyhq/twenty/blob/main/packages/twenty-server/src/modules/messaging/services/fetch-messages-by-batches/fetch-messages-by-batches.service.ts#L70-L74

>Can we infer a type from gmail_v1.Schema$Message with all fields we need required?

I mean, this is the best type we can get, or is it not? It comes straight from Google that clearly says some properties might are optional. When testing, those we need were always defined, but isnt it better to be safe than sorry? Im doing a runtime assertion using:
```
assert(id);
assert(threadId);
assert(historyId);
assert(internalDate);
```
This would throw early (message wouldnt be parsed) if some of them are actually missing. Safer approach actually rather than letting it throw further down the line with "cant read property X of undefined" or saving invalid data into the DB.

Let me know what you think about these two topics :)

File: .vscode/launch.json (hunk 1 above)
Comment (rostaklein) [reply .vscode/launch.json:38]: Nice! It finally works! :) The missing peace for me was actually uncommenting the `MESSAGE_QUEUE_TYPE` env variable so its not doing the sync type of queue which always landed in one process and the worker didnt pick it up :D Im still learning a lot about this project and there are always some parts which you guys take for granted and im getting surprised :)

Comment (rostaklein) [reply .vscode/launch.json]: + i got finally a person created out of the conversation as well 🎉  had to figure out all these conditions that were made with a good intention 😄 (sending a gmail to gmail message didnt do what I wanted it to do)

File: packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts (hunk 2 above)
Comment (charlesBochet) [reply packages/twenty-server/src/modules/messaging/types/gmail-message-parsed-response.ts:3]: I prefer the assert pattern that you are using

I've just remove the errors from your type and it looks good

**Comment by charlesBochet** [comment]
@rostaklein I've rebased your PR on main and made minor changes.

Now I have two questions:
1) why did we stop fetching participants from cc and bcc in fetch-messages-by-batch.service.ts?
2) I have tried to import my email. And for some messages, "Message-ID" header is actually "Message-Id". I've added toLowerCase in the header check ==> I could not fully test that this was functionnal because I'm hitting my Google API dev rate limiting. I'll re-try in 1 hour

**Comment by charlesBochet** [comment]
I was able to re-test and it seems to work.
However, you are right the error handling is broken. Here is what we should do IMO:
- very early (so in fetchAllMessages from FetchMessagesByBatchesService) we should understand if we have an error or not. If we have a 404, let's just drop the message. if we have anything else (429, 403, 400, 500) we should put the message channel syncStatus in FAILED state (we will revisit these status later, they are not precise enough)

We should stop handling an array of errors in the code ; either we ignore for 404 and just drop the message, either we throw and put the whole channel in failed state. Treating these error edge cases early is the best way to have a simple and robust code. Anyway if we get a 429, 403... on a message, we will get it for all messages

**Comment by rostaklein** [comment]
>why did we stop fetching participants from cc and bcc in fetch-messages-by-batch.service.ts

For some reason I thought the new format is not returning it. I was mistaken. It is there indeed so Ive added it back 😉 

>404: silent + remove the message from the list / other error code: throw

Good idea. Done. 👍 Not sure how much youll like the solution thoug. I had to do it this way 👇 as thats exactly what we are getting via axios from FetchByBatchesService
```ts
export type GmailMessageParsedResponse =
  | gmail_v1.Schema$Message
  | GmailMessageError;
 ```

File: packages/twenty-server/src/modules/messaging/services/gmail-fetch-message-content-from-cache/gmail-fetch-message-content-from-cache.service.ts [hunk 3]
```diff
@@ -292,21 +276,19 @@ export class GmailFetchMessageContentFromCacheService {
...
 
-      if (error?.message?.code === 429) {
-        this.logger.error(
-          `Error fetching messages for ${connectedAccountId} in workspace ${workspaceId}: Resource has been exhausted, locking for ${GMAIL_ONGOING_SYNC_TIMEOUT}ms...`,
-        );
+      await this.messageChannelRepository.updateSyncStatus(
+        gmailMessageChannelId,
+        MessageChannelSyncStatus.FAILED,
+        workspaceId,
+      );
```
Comment (rostaklein) [inline packages/twenty-server/src/modules/messaging/services/gmail-fetch-message-content-from-cache/gmail-fetch-message-content-from-cache.service.ts:283]: This is now being called for every single caught error, not just 429. I think thats what we actually want, right?

Comment (charlesBochet) [reply packages/twenty-server/src/modules/messaging/services/gmail-fetch-message-content-from-cache/gmail-fetch-message-content-from-cache.service.ts:283]: Yes!

**Comment by charlesBochet** [comment]
I've looked at your code, that's it!
I'll test it with all edge cases tonight

**Comment by rostaklein** [comment]
Let me know whether theres anything else to be done, or we can continue with the further changes..? :)

**Comment by charlesBochet** [comment]
@rostaklein Sorry, it took a long time to tests all edge cases. I've pushed a few fixes, it's working well! We can move to the attachment logic!

**Comment by charlesBochet** [comment]
We are going to revamp the jobs (but not the logic you've changed), according to:
image
 This will improve the stability

**Review by charlesBochet — APPROVED** [review]
LGTM, thank you!
